File: python/data_structures/treap.py
"""this is the implementation of a treap class
https://en.wikipedia.org/wiki/Treap
the ADT contains the following methods:
    - insert
    - search
    - delete
"""
from data_structures import bst
from random import uniform

class Treap(bst.BST):
    class Node():
        """Node basic chainable storage unit
        """
        def __init__(self, value=None):
            self.value = value
            self.priority = uniform(0,1)
            self.left = None
            self.right = None

        def __repr__(self):
            return repr(self.value)

    def __init__(self, x=[]):
        self.root = None
        for e in x:
            self.insert(e)

    def rotate_right(self, node):
        """right rotation...
        https://en.wikipedia.org/wiki/AVL_tree#Simple_rotation
        """
        tmp = self.Node(node.value)
        tmp.priority = node.priority
        tmp.left, tmp.right = node.left.right, node.right
        # rotate...
        node.value, node.priority = node.left.value, node.left.priority
        node.left, node.right = node.left.left, tmp

    def rotate_left(self, node):
        """left rotation...
        https://en.wikipedia.org/wiki/AVL_tree#Simple_rotation
        """
        tmp = self.Node(node.value)
        tmp.priority = node.priority
        tmp.left, tmp.right = node.left, node.right.left
        # rotate...
        node.value, node.priority = node.right.value, node.right.priority
        node.left, node.right = tmp, node.right.right

    def insert(self, value):
        """insert a new node with the given value in the tree
        standard bst insertion + rotations to keep the heap invariance
        """
        if value == None: return
        if not self.root: self.root = self.Node(value); return
        # execute standard bst insertion...
        def r(node):
            if value == node.value: raise ValueError(f'No duplicates allowed...')
            if value < node.value:
                if node.left: r(node.left)
                else: node.left = self.Node(value)
            else:
                if node.right: r(node.right)
                else: node.right = self.Node(value)
            # repair max_heap invariants...
            if value > node.value:
                if node.right.priority > node.priority:
                    self.rotate_left(node)
            else:
                if node.left.priority > node.priority:
                    self.rotate_right(node)
        r(self.root)

    def delete(self, value):
        """search and delete the node with the given value in the tree
        rotate the node to be deleted with the max child until if becomes a leaf
        then just erase the leaf...
        """
        if value == None: return
        # standard binary search for the node, and the parent
        if not self.root: raise ValueError(f'{value} not found')
        def get(node, parent):
            if not node: raise ValueError(f'{value} not found')
            if value == node.value: return (node, parent)
            if value < node.value: return get(node.left, node)
            else: return get(node.right, node)
        node, parent = get(self.root, None)
        # make it a leaf so we can get rid of it easily
        def r(node, parent):
            # base case: node is a leaf
            if not node.left and not node.right:
                # are you root?
                if not parent: self.root = None; return
                if parent.right == node: parent.right = None; return
                else: parent.left = None; return
            # rotate with max_child, until no childs remain...
            priority = lambda node: node.priority if node else -float('inf')
            max_child = max(node.left, node.right, key=lambda node: priority(node))
            if max_child == node.right:
                self.rotate_left(node)
                return r(node.left, node)
            else:
                self.rotate_right(node)
                return r(node.right, node)
        return r(node, parent)
    
    def search(self, value):
        """standard binary search in the tree
        """
        if not self.root: raise ValueError(f'{value} not found')
        def r(node):
            if not node: raise ValueError(f'{value} not found')
            if value == node.value: return node
            if value < node.value: return r(node.left)
            else: return r(node.right)
        return r(self.root)

    def __repr__(self):
        if not self.root: return
        res = []
        def r(node, level=0):
            if not node: return
            r(node.left, level +1)
            res.append('\t'*level + f'-->({node.value}){node.priority:0.04}')
            r(node.right, level +1)
        r(self.root)
        return '\n'.join(res)

File: cpp/src/treap/src/treap.cpp
#include "treap/treap.h"

namespace treap
{

template<typename T>
Node<T>::Node(T _value): value(_value) {
	this->priority = this->uniform();
}

template<typename T>
double Node<T>::uniform() {
	std::random_device rd;  //Will be used to obtain a seed for the random number engine
    std::mt19937 gen(rd()); //Standard mersenne_twister_engine seeded with rd()
	std::uniform_real_distribution<double> u(0, 1);
	return u(gen);
}

template<typename T>
Treap<T>::Treap() {}

template<typename T>
Treap<T>::Treap(const std::vector<T>& data) {
	for(T e: data)
		this->insert(e);
}

template<typename T>
void Treap<T>::insert(const T& key) {
	if(!this->root) {
		this->root = std::shared_ptr<Node<T>>(new Node<T>(key));
		return;
	}
	return this->insert(key, this->root);
}

template<typename T>
void Treap<T>::insert(const T& key, std::shared_ptr<Node<T>> node) {
	if(!node)
		return;
	//standard insertion
	if(node->value == key)
		throw std::invalid_argument("No, duplicates allowed!");
	if(key < node->value) {
		if(node->left) {
			this->insert(key, node->left);
		}
		else
			node->left = std::shared_ptr<Node<T>>(new Node<T>(key));
		//repair violations
		if(node->priority < node->left->priority)
			this->rotate_right(node);
	}
	else {
		if(node->right) {
			this->insert(key, node->right);
		}
		else
			node->right = std::shared_ptr<Node<T>>(new Node<T>(key));
		//repair violations
		if(node->priority < node->right->priority)
			this->rotate_left(node);
	}
}

template<typename T>
std::shared_ptr<Node<T>> Treap<T>::max() {
	if(!this->root)
		return nullptr;
	auto node = this->root;
	while(node->right)
		node = node->right;
	return node;
}

template<typename T>
std::shared_ptr<Node<T>> Treap<T>::min() {
	if(!this->root)
		return nullptr;
	auto node = this->root;
	while(node->left)
		node = node->left;
	return node;
}

template<typename T>
std::shared_ptr<Node<T>> Treap<T>::successor(const T& key) {
	if(!this->root)
		return nullptr;
	return this->successor(key, this->root);
}

template<typename T>
std::shared_ptr<Node<T>> Treap<T>::successor(const T& key, 
										   std::shared_ptr<Node<T>> node,
										   std::shared_ptr<Node<T>> ancestor) {
	if(!node)
		return ancestor;
	if(node->value == key) {
		//no right child to explore, return the next biggest ancestor
		if(!node->right)
			return ancestor;
		//explore and return the min of the left subtree
		else {
			auto tmp = node->right;
			while(tmp->left)
				tmp = tmp->left;
			return tmp;
		}
	}
	if(key < node->value)
		return this->successor(key, node->left, node);
	else
		return this->successor(key, node->right, ancestor);
}

/// standard binary search
template<typename T>
std::shared_ptr<Node<T>> Treap<T>::search(const T& key) {
	if(!this->root)
		return nullptr;
	return this->search(key, this->root);
}

template<typename T>
std::shared_ptr<Node<T>> Treap<T>::search(const T& key,
										std::shared_ptr<Node<T>> node) {
	if(!node)
		return nullptr;
	if(key == node->value)
		return node;
	if(key < node->value)
		return this->search(key, node->left);
	else
		return this->search(key, node->right);
}

/// standard right rotation: https://en.wikipedia.org/wiki/Treap_tree#Simple_rotation
template<typename T>
void Treap<T>::rotate_right(std::shared_ptr<Node<T>> node) {
	if(!node)
		return;
	//set-up
	auto tmp = std::shared_ptr<Node<T>>(new Node<T>(node->value));
	tmp->priority = node->priority;
	tmp->left = node->left->right;
	tmp->right = node->right;
	//rotate
	node->value = node->left->value;
	node->priority = node->left->priority;
	node->left = node->left->left;
	node->right = tmp;
}

/// standard left rotation: https://en.wikipedia.org/wiki/Treap_tree#Simple_rotation
template<typename T>
void Treap<T>::rotate_left(std::shared_ptr<Node<T>> node) {
	if(!node)
		return;
	//set-up
	auto tmp = std::shared_ptr<Node<T>>(new Node<T>(node->value));
	tmp->priority = node->priority;
	tmp->left = node->left;
	tmp->right = node->right->left;
	//rotate
	node->value = node->right->value;
	node->priority = node->right->priority;
	node->left = tmp;
	node->right = node->right->right;
}

/// initializer for the recursive erase function
template<typename T>
void Treap<T>::erase(const T& key) {
	if(!this->root)
		return;
	std::shared_ptr<Node<T>> father = nullptr;
	return erase(key, this->root, father);
}

/// rotate the node to be erased until it becomes a leaf and then delete is trivial
template<typename T>
void Treap<T>::erase(const T& key,
				   std::shared_ptr<Node<T>> node,
				   std::shared_ptr<Node<T>> father) {
	if(!node)
		return;
	if(node->value == key) {
		//have children? keep rotating...
		if(node->left || node->right) {
			auto child = (node->left)?node->left:node->right;
			if(node->left == child) {
				this->rotate_right(node);
				this->erase(key, node->right, node);
			}
			else {
				this->rotate_left(node);
				this->erase(key, node->left, node);
			}
		}
		//no children => erase the leaf easily
		else {
			//special case: are you root?
			if(!father){
				this->root = nullptr;
				return;
			}
			//general case:
			if(father->left == node)
				father->left = nullptr;
			else
				father->right = nullptr;
			return;
		}
	}
	//keep looking
	else {
		if(key < node->value) 
			this->erase(key, node->left, node);
		else
			this->erase(key, node->right, node);
	}
}

template<typename T>
void Treap<T>::traverse() {
	if(!this->root)
		return;
	return traverse(this->root);
}

template<typename T>
void Treap<T>::traverse(std::shared_ptr<Node<T>> node, int level) {
	if(!node)
		return;
	this->traverse(node->left, level+1);
	for(auto i = 0; i < level; i++)
		std::cout << "\t";
	std::cout << "-->(" << std::to_string(node->value) << ")"
				<< "p" << node->priority << std::endl;
	this->traverse(node->right, level+1);
}

} // namespace treap

// only here for testing while we implement the gtest module
int main()
{
	std::vector<int> test = {5,7,0,9,3,2,8,12,1};
	auto tree = treap::Treap<int>(test);
	tree.traverse();
	auto to_erase = 5;
	std::cout << "----- erasing: " << to_erase << " -----" << std::endl;
	tree.erase(to_erase);
	tree.traverse();
	auto node = tree.search(9);
	if(node)
		std::cout << "node found: " << std::to_string(node->value) << std::endl;
	node = tree.max();
	if(node)
		std::cout << "max node found: " << std::to_string(node->value) << std::endl;
	node = tree.min();
	if(node)
		std::cout << "min node found: " << std::to_string(node->value) << std::endl;
	node = tree.successor(8);
	if(node)
		std::cout << "successor node found: " << std::to_string(node->value) << std::endl;
	return 0;
}
